feature/generate order_items based on forecast

elijahmylcev opened the pull request:
## **Type**
enhancement

___

## **Description**
- Added a new Alembic migration to modify the 'order_item' table by replacing 'status' with 'status_id'.
- Scheduled automated tasks for generating orders based on forecasts using BackgroundScheduler.
- Implemented classes and methods to calculate and generate orders based on forecast data.
- Integrated with a forecasting service to fetch forecast data.
- Updated models and added new Pydantic serializers for handling and validating data.


___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>2024_04_22_1340-17af7f7b8bd1_order_items_have_status_id.py</strong><dd><code>Add status_id to order_item and adjust constraints</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

alembic/versions/2024_04_22_1340-17af7f7b8bd1_order_items_have_status_id.py
<li>Added a new Alembic migration to modify the 'order_item' table by <br>replacing 'status' with 'status_id'.<br> <li> Added and dropped constraints related to the 'status_id' changes.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-d9d6d4b1eea32b5bc69cb15af0e5d58fb48f525cc6cda5894ccf8af63b8eda8b">+35/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>server.py</strong><dd><code>Schedule automated order generation tasks</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/configuration/server.py
<li>Added automated task scheduling for generating orders based on <br>forecasts.<br> <li> Configured a BackgroundScheduler to run the order generation task <br>periodically.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-7ea74dc5643513861f4947fdb882390b47017bcc3a174ea78163f739d7be7e5c">+8/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>calculate_orders.py</strong><dd><code>Implement order calculation from forecast data</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/domain/calculate_orders.py
<li>Implemented a new class to calculate orders based on forecast data.<br> <li> Handles timezone adjustments and multiplies amounts by ingredient <br>coefficients.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-1c38d58864e20337a9bed95c95054dc32da486e5e7dc98d7ff97fe33e60260f9">+53/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>generate_orders.py</strong><dd><code>Implement event to generate orders from forecast</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/internal/events/generate_orders.py
<li>New function to fetch forecast and generate orders using the <br>CalculateOrders class.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-a654fe0a859a64028c9218110fad7d7e0c5b979196b1e5fb0affe3b7d5d0507a">+18/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>get_forecast.py</strong><dd><code>Integrate with forecasting service to fetch data</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/internal/integrations/get_forecast.py
<li>New class to integrate with a forecasting service and fetch forecast <br>data.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-5e8c0637cfd8b58416cddcb426944e4b497dac9c6c13b260314af07a972a4ce2">+20/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>order_item.py</strong><dd><code>Update OrderItem model and add data handling methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/models/order_item.py
<li>Updated OrderItem model to support new 'status_id' field.<br> <li> Added methods to create or update orders based on new or existing <br>data.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-4499f75bab6a767750030849dd39f4642489a63bdbb1497c6b0df8eaaa4f0141">+60/-6</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>product_ingredient_association.py</strong><dd><code>Enhance ProductIngredientAssociation model with new data fetching </code><br><code>method</code></dd></summary>
<hr>

app/models/product_ingredient_association.py
<li>Added method to fetch recipe items by product ID.<br> <li> Integrated with session management for database operations.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-ff2085093784b34238bdfa734e7bf28b442e702fcc122616da2bdacedca8c7ac">+68/-2</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ingredient.py</strong><dd><code>Define Pydantic model for ingredient serialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/serializers/ingredient.py
- New Pydantic model for ingredient serialization.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-f09114eee6fd807c1000e5c71b8ff7b93ef594ad06205b2ebc974231c97bdac2">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>product.py</strong><dd><code>Define Pydantic model for product serialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/serializers/product.py
- New Pydantic model for product serialization.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-4d9e563180b605202f59f2ec7e4273654dc7b27c786059b0b92ebdbfa9323c41">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>product_ingredient_association.py</strong><dd><code>Define Pydantic model for product ingredient association serialization</code></dd></summary>
<hr>

app/serializers/product_ingredient_association.py
<li>New Pydantic model for product ingredient association serialization.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-b3b2e41f10843573d46c984e76fc7c25e97d0747f67ae1b2431ef451a76ebf6f">+12/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>units_of_measurement.py</strong><dd><code>Define Pydantic model for unit of measurement serialization</code></dd></summary>
<hr>

app/serializers/units_of_measurement.py
- New Pydantic model for unit of measurement serialization.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-07f6d8c0c69bf98fe95522b5732963c9ec1d2f63b49503fe595cdca55673e658">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>settings.py</strong><dd><code>Add new environment variable for forecast URL</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

app/configuration/settings.py
<li>Added a new environment variable URL_FORCAST to fetch forecast data.


</details>
    

  </td>
  <td><a href="https://github.com/farm-service/backend/pull/16/files#diff-18cdd0e8d458a1fd8b4b10615c5e81cdbdd25455c8a89c9008f483955ef0da52">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

elijahmylcev committed:
feature/generate order_items based on forecast

elijahmylcev committed:
feature/generate order_items; added Logger

elijahmylcev committed:
feature/generate order_items; freeze requirements.txt

elijahmylcev committed:
feature/generate order_items; SQLAlchemy==1.* requirements.txt

Tardigrada777 commented on app/configuration/server.py:
```diff
@@ -30,7 +33,14 @@ def get_app(self) -> FastAPI:
     @staticmethod
     def __register_events(app: FastAPI) -> None:
-        ...
+        app.on_event('startup')(repeat_every(seconds=60 * 60 * 12)(generate_orders))
```
According to the documentation https://github.com/dmontagu/fastapi-utils/blob/master/fastapi_utils/tasks.py#L26 `repet_every` function:

> This function returns a decorator that modifies a function so it is periodically re-executed after its first call.

This means the `generate_orders` event will be repeated every 12 hours (according to the presented configuration). It's probably not we are expecting here.

elijahmylcev replied:
Fixed

Tardigrada777 commented on app/domain/calculate_orders.py:
```diff
@@ -0,0 +1,55 @@
+from datetime import datetime, timedelta
+
+import pytz
+from sqlalchemy.ext.declarative import DeclarativeMeta
```
Business logic code from the `domain` level is not allowed to import anything from the `db` layer.

Tardigrada777 commented on app/domain/calculate_orders.py:
```diff
@@ -0,0 +1,55 @@
+from sqlalchemy.ext.declarative import DeclarativeMeta
+from typing import List, Optional, Dict, Type
+
+from app.configuration.settings import logger
```
Business logic code from the `domain` level is not allowed to import anything from the `application` layer. Logger - is an application layer entity.

Tardigrada777 commented on app/domain/calculate_orders.py:
```diff
@@ -0,0 +1,55 @@
+                        )
+            return result
+        except Exception as e:
+            logger.error(f'Exception occurred: {e}')
```
You can handle this exception on the layer above (e.g. on the application layer, where we describe the event code).

Tardigrada777 commented on app/domain/calculate_orders.py:
```diff
@@ -0,0 +1,55 @@
+    async def process(
+            cls,
+            forecast: Optional[Dict[str, dict]],
+            recipe_model: Type[DeclarativeMeta],
```
You can use generic types here in order to avoid direct import of `db` layer entities.

elijahmylcev committed:
feature/generate order_items; fixed requirements.txt and domain logic

Tardigrada777 commented on app/configuration/server.py:
```diff
@@ -30,7 +32,14 @@ def get_app(self) -> FastAPI:
     @staticmethod
     def __register_events(app: FastAPI) -> None:
-        ...
+        app.on_event('startup')(generate_orders)
```
Do we really need to run this event on `startup`? Or it's enough to schedule it as a cron event?

Tardigrada777 commented on app/domain/calculate_orders.py:
```diff
@@ -0,0 +1,52 @@
+    async def process(
+            cls,
+            forecast: Optional[Dict[str, dict]],
+            recipe_model: Generic[T],
```
This called `injection`: when we inject an object in runtime to function or method in order to use object in runtime.

It's probably OK for this specific case (I would go with constructor-based injection though). But let's try at least define types for models.

It's not OK that domain logic uses word "model" from the `db` layer.

We should rather:

1. Define type for repository - object with access to DB
2. Make `process` method depend on the type, not actual implementation details
3. Inject instance of repository class (that may use models inside) on the layer above

elijahmylcev committed:
feature/generate order_items; fixed domain_logic

Tardigrada777 commented:
/improve

elijahmylcev committed:
feature/generate order_items; Comments have been corrected

elijahmylcev commented:
@Tardigrada777 I have corrected all comments. All that remains is to fix the test (@ChaPerx64). I'm waiting for your comments!

ChaPerx64 committed:
cd: docker-test robustness improved
curl now retries requests every 5 seconds for 2 minutes
docker logs are output in case of failure

ChaPerx64 committed:
Merge branch 'main' into feature/geneate_orders_from_forecast

ChaPerx64 committed:
fix: docker-test, echo command added

ChaPerx64 committed:
feat(cd): error message added for enchanced readability

Slautin1703 commented on app/domain/repositories/calculate_orders_repository_interface.py:
```diff
@@ -0,0 +1,32 @@
+            deadline: datetime,
+            status_id: int,
+            ingredient_id: int,
+            consumer_id: any,
```
@elijahmylcev i would kindly reccomend to use specific type for this parameter

Slautin1703 commented on app/internal/events/calculate_orders_repository.py:
```diff
@@ -0,0 +1,42 @@
+        Returns:
+            Optional[List[ABC]]: List of recipes by product_id or None
+        """
+        return await ProductIngredientAssociation.get_items_by_product(product_id)
```
as i think its better to return consistent data. So if there is no items - just return empty array
